tests/main.rs

```diff
@@ -112,26 +141,78 @@ where
         self
     }
 
-    #[cfg(feature = "tls")]
-    fn with_server_tls(mut self) -> Self {
-        let cert = rcgen::generate_simple_self_signed(vec!["localhost".to_string()]).unwrap();
-
-        self.server_tls = Some(std::sync::Arc::new(
-            ServerConfig::builder()
-                .with_safe_defaults()
-                .with_no_client_auth()
-                .with_single_cert(
-                    vec![Certificate(cert.serialize_der().unwrap())],
-                    PrivateKey(cert.get_key_pair().serialize_der()),
-                )
-                .unwrap(),
-        ));
+    fn with_tls(mut self, client: bool, server: bool, use_client_certs: bool) -> Self {
+        #[cfg(feature = "tls")]
+        let mut client_cert_der = None;
 
-        self
-    }
+        if use_client_certs {
+            let (client_cert, client_pkey) = mk_client_cert().unwrap();
+
+            #[cfg(feature = "tls")]
+            {
+                client_cert_der = Some(Certificate(client_cert.to_der().unwrap()));
+            }
+
+            // Set up client cert der12 file.
+            let client_cert_pkcs12_file = Arc::new(tempfile::NamedTempFile::new().unwrap());
+            self.client_cert_pkcs12_file = Some(Arc::clone(&client_cert_pkcs12_file));
+
+            let pkcs12 = Pkcs12::builder()
+                .build("password", "friendly_name", &client_pkey, &client_cert)
+                .unwrap();
+            let der = pkcs12.to_der().unwrap();
+
+            let mut f = File::create(&*client_cert_pkcs12_file).unwrap();
+            f.write(&der).unwrap();
+            f.flush().unwrap();
+        }
+
+        #[cfg(not(feature = "tls"))]
+        if server {
+            panic!("a test is asking for tls support in the server, but we don't have the tls feature enabled.");
+        }
+
+        #[cfg(feature = "tls")]
+        if server {
+            let cert = generate_simple_self_signed(vec!["localhost".to_string()]).unwrap();
+
+            let builder = ServerConfig::builder().with_safe_defaults();
+
+            let builder = if use_client_certs {
+                let mut client_auth_roots = RootCertStore::empty();
+
+                client_auth_roots.add(&client_cert_der.unwrap()).unwrap();
+
+                let client_auth = AllowAnyAuthenticatedClient::new(client_auth_roots);
+
+                builder.with_client_cert_verifier(client_auth)
+            } else {
+                builder.with_no_client_auth()
+            }
+            .with_single_cert(
+                vec![Certificate(cert.serialize_der().unwrap())],
+                PrivateKey(cert.get_key_pair().serialize_der()),
+            )
+            .unwrap();
+
+            self.server_tls = Some(std::sync::Arc::new(builder));
+        }
+
+        if client {
+            let mut opts = SslOpts::default().with_danger_accept_invalid_certs(true);
+
+            if use_client_certs {
```

Comment:
I don't think you need the contained code to be conditional on `use_client_certs`, since `self.client_cert_pkcs12_file` will be `None` anyway if it's `false`.